File: src/examples/ex_table_pour.cpp
/* ex_table_pour.cpp
 * Kinova GEN3 + Robotiq 2F-85 pours small balls from a small attached bottle into a
 * transparent tabletop receiver, with joint impedance (PD + KDL gravity) in TORQUE mode.
 *
 * Usage:
 *   ex_table_pour [--headless] [--record output.mp4]
 *
 * Runs the pour sequence once and prints how many balls ended in the receiver; --headless skips
 * the viewer and exits 1 if fewer than kMinInReceiver did; --record writes an MP4 offscreen
 * (EGL + ffmpeg) and implies --headless. */

#include "common.hpp"
#include "example_paths.hpp"

#include <kdl/chainiksolverpos_lma.hpp>
#include <kdl/chainiksolverpos_nr_jl.hpp>
#include <kdl/chainiksolvervel_pinv.hpp>

#include <algorithm>
#include <cmath>
#include <cstdio>
#include <iomanip>
#include <iostream>
#include <string>
#include <vector>

namespace ex = mj_kdl_examples;

static constexpr double kRobotBackX    = -0.26;
static constexpr double kJugX          = 0.30;
static constexpr double kJugY          = 0.14;
static constexpr double kRetreatX      = kJugX - 0.08;
static constexpr double kRetreatY      = kJugY - 0.08;
static constexpr double kJugRadius     = 0.028;
static constexpr double kJugHeight     = 0.084;
static constexpr int    kNumBalls      = 36;
static constexpr double kBallRadius    = 0.007;
static constexpr double kIkTol         = 3e-3;
static constexpr double kPourTiltRad   = 1.95;
static constexpr double kTiltOutletZ   = ex::kTableZ + 0.18;
static constexpr int    kMinInReceiver = 24;
static constexpr int    kRecordFps     = 60;

static constexpr double kKp[7] = { 120, 220, 120, 220, 110, 190, 90 };
static constexpr double kKd[7] = { 12, 22, 12, 22, 11, 18, 9 };

static mj_kdl::SceneObject make_ball(int idx)
{
    mj_kdl::SceneObject ball;
    char                name[32];
    std::snprintf(name, sizeof(name), "grain_%02d", idx);
    ball.name    = name;
    ball.shape   = mj_kdl::Shape::SPHERE;
    ball.size[0] = kBallRadius;
    ball.size[1] = ball.size[2] = 0.0;
    ball.pos[2]                 = ex::kTableZ + 0.40 + idx * 2.0 * kBallRadius;
    const float rgba[4]         = { 1.0f, 0.84f, 0.30f, 1.0f };
    std::copy(rgba, rgba + 4, ball.rgba);
    ball.mass        = 0.006;
    ball.condim      = mj_kdl::Condim::Torsional;
    ball.friction[0] = 0.5;
    ball.friction[1] = 0.02;
    ball.friction[2] = 0.001;
    return ball;
}

static bool inside_receiver(const KDL::Vector &p)
{
    return std::abs(p.x() - kJugX) < (kJugRadius - 0.012)
           && std::abs(p.y() - kJugY) < (kJugRadius - 0.012) && p.z() > ex::kTableZ + 0.006
           && p.z() < ex::kTableZ + kJugHeight + 0.04;
}

int main(int argc, char *argv[])
{
    const ex::Args args     = ex::parse_args(argc, argv, "table_pour.mp4");
    const bool     headless = args.headless;

    mj_kdl::AttachmentSpec bottle;
    bottle.mjcf_path = ex::asset("mug.xml");
    bottle.attach_to = { mj_kdl::AttachKind::Body, "g_base" };
    bottle.prefix    = "pour_";

    mj_kdl::RobotSpec robot_spec;
    robot_spec.path   = ex::menagerie_model("kinova_gen3/gen3.xml");
    robot_spec.pos[0] = kRobotBackX;
    robot_spec.pos[2] = ex::kTableZ;
    robot_spec.attachments.push_back(ex::gripper_attachment(ex::asset("robotiq_2f85/2f85.xml")));
    robot_spec.attachments.push_back(bottle);

    mj_kdl::SceneSpec scene_cfg = ex::scene_spec();
    scene_cfg.objects.push_back(ex::table_object(ex::asset("table.xml"), ex::kTableZ));
    for (int i = 0; i < kNumBalls; ++i) scene_cfg.objects.push_back(make_ball(i));
    mj_kdl::SceneObject receiver;
    receiver.name      = "recv";
    receiver.mjcf_path = ex::asset("mug_table.xml");
    receiver.pos[0]    = kJugX;
    receiver.pos[1]    = kJugY;
    receiver.pos[2]    = ex::kTableZ;
    scene_cfg.objects.push_back(receiver);
    scene_cfg.robots.push_back(robot_spec);

    mj_kdl::Env   env;
    mj_kdl::Robot robot;
    if (!mj_kdl::init_env(&env, &scene_cfg)) return 1;
    const mj_kdl::ToolFrameSpec tool = ex::gripper_tool();
    if (!mj_kdl::init_robot_from_mjcf(&robot, &env, "base_link", "bracelet_link", "", &tool))
        return 1;
    if (!mj_kdl::set_control_mode(&robot, mj_kdl::CtrlMode::TORQUE)) return 1;
    mj_kdl::SceneActuatorSlot *fingers =
      mj_kdl::bind_scene_actuator(&env.scene, "g_fingers_actuator");
    if (!fingers) return 1;
    std::vector<mj_kdl::SceneFreeBodySlot *> grains;
    for (int i = 0; i < kNumBalls; ++i) {
        char name[32];
        std::snprintf(name, sizeof(name), "grain_%02d", i);
        grains.push_back(mj_kdl::bind_scene_free_body(&env.scene, name));
        if (!grains.back()) return 1;
    }

    const unsigned                  n      = robot.chain.getNrOfJoints();
    const KDL::JntArray             q_home = ex::home_q(n);
    KDL::ChainFkSolverPos_recursive fk(robot.chain);
    KDL::JntArray                   q_min(n), q_max(n);
    for (unsigned i = 0; i < n; ++i) {
        const auto [lo, hi] = robot.joint_limits[i];
        q_min(i)            = std::isfinite(lo) ? lo : -2 * M_PI;
        q_max(i)            = std::isfinite(hi) ? hi : 2 * M_PI;
    }
    KDL::ChainIkSolverVel_pinv  ik_vel(robot.chain);
    KDL::ChainIkSolverPos_NR_JL ik_nr(robot.chain, q_min, q_max, fk, ik_vel, 2000, 1e-5);
    KDL::ChainIkSolverPos_LMA   ik_lma(robot.chain, 1e-5, 2000);
    KDL::ChainDynParam          dyn(robot.chain, KDL::Vector(0.0, 0.0, scene_cfg.gravity_z));

    KDL::Frame home_fk;
    fk.JntToCart(q_home, home_fk);
    const KDL::Rotation carry_tcp = home_fk.M * KDL::Rotation::RotY(-0.05);
    const KDL::Frame    base_T_world =
      KDL::Frame(KDL::Vector(kRobotBackX, 0.0, ex::kTableZ)).Inverse();

    // The TCP -> outlet offset, measured at home, turns an outlet target into a TCP target.
    mj_kdl::set_joint_pos(&robot, q_home);
    KDL::Frame world_T_outlet, world_T_tcp;
    mj_kdl::get_site_frame(&env, "pour_outlet", &world_T_outlet);
    mj_kdl::get_site_frame(&env, "g_pinch", &world_T_tcp);
    const KDL::Vector tcp_outlet = world_T_tcp.Inverse() * world_T_outlet.p;

    const auto solve =
      [&](
        const char *name, const KDL::Vector &outlet, const KDL::JntArray &seed, KDL::JntArray &out
      ) {
          const KDL::Frame target =
            base_T_world * KDL::Frame(carry_tcp, outlet - carry_tcp * tcp_outlet);
          bool ok = ik_nr.CartToJnt(seed, target, out) >= 0;
          if (!ok) ok = ik_lma.CartToJnt(seed, target, out) >= 0;
          KDL::Frame fk_out;
          fk.JntToCart(out, fk_out);
          if (!ok || (target.p - fk_out.p).Norm() > kIkTol) {
              std::cerr << "IK failed for " << name << "\n";
              return false;
          }
          return true;
      };

    KDL::JntArray q_pre_pour(n), q_pour(n), q_tilt(n), q_retreat(n);
    KDL::Vector   pour_outlet(kJugX, kJugY, ex::kTableZ + 0.20);
    if (!solve("pre-pour", KDL::Vector(kJugX, kJugY, ex::kTableZ + 0.27), q_home, q_pre_pour))
        return 1;
    if (!solve("pour", pour_outlet, q_pre_pour, q_pour)) return 1;
    // Tilting swings the outlet away; shift the pour pose until the tilted outlet is on target.
    for (int iter = 0; iter < 4; ++iter) {
        q_tilt = q_pour;
        q_tilt(n - 1) += kPourTiltRad;
        mj_kdl::set_joint_pos(&robot, q_tilt);
        mj_kdl::get_site_frame(&env, "pour_outlet", &world_T_outlet);
        const KDL::Vector err = KDL::Vector(kJugX, kJugY, kTiltOutletZ) - world_T_outlet.p;
        if (err.Norm() < 5e-3) break;
        pour_outlet += err;
        if (!solve("pour", pour_outlet, q_pre_pour, q_pour)) return 1;
    }
    q_tilt = q_pour;
    q_tilt(n - 1) += kPourTiltRad;
    if (!solve("retreat", KDL::Vector(kRetreatX, kRetreatY, ex::kTableZ + 0.27), q_pour, q_retreat))
        return 1;

    bool restart = false;
    env.on_reset = [&](mj_kdl::ResetContext *ctx) {
        mj_kdl::set_joint_pos(&robot, q_home);
        KDL::Frame world_T_center;
        mj_kdl::get_site_frame(&env, "pour_center", &world_T_center);
        const double spacing = 2.0 * kBallRadius;
        for (int i = 0; i < kNumBalls; ++i) {
            const int         layer = i / 9, slot = i % 9;
            const KDL::Vector local(
              (slot % 3 - 1) * spacing, (slot / 3 - 1) * spacing, -0.026 + layer * spacing
            );
            const KDL::Vector w      = world_T_center * local;
            const double      pos[3] = { w.x(), w.y(), w.z() };
            mj_kdl::set_body_pose(&env, grains[i]->name.c_str(), pos);
        }
        ctx->data->ctrl[fingers->ctrl_id] = ex::kGripperClosed;
        ex::prime_gravity(robot, dyn, q_home);
        restart = true;
    };

    const double close = ex::kGripperClosed;
    // clang-format off
    const std::vector<ex::Phase> phases = {
        { "HOME",      &q_home,     1.0,  2.5,  0.08, close },
        { "PRE_POUR",  &q_pre_pour, 4.0,  6.5,  0.08, close },
        { "POUR",      &q_pour,     3.5,  5.5,  0.07, close },
        { "TILT",      &q_tilt,     7.0, 10.0,  0.07, close },
        { "POUR_HOLD", &q_tilt,     9.0,  9.0, -1.0,  close },
        { "RETREAT",   &q_retreat,  2.0,  4.0,  0.08, close },
        { "HOLD",      &q_retreat,  1.0,  1.0, -1.0,  close },
    };
    // clang-format on

    mj_kdl::VideoRecorder recorder;
    bool                  recording = false;
    const int             steps_per_frame =
      std::max(1, static_cast<int>(std::lround(1.0 / (kRecordFps * env.model->opt.timestep))));
    if (args.record) {
        const mj_kdl::Status s = mj_kdl::init_video_recorder(
          &recorder,
          env.model,
          args.record_path.c_str(),
          mj_kdl::VideoResolution::R1080p,
          kRecordFps
        );
        if (!s) {
            std::cerr << "init_video_recorder() failed: " << s.error << "\n";
            return 1;
        }
        recorder.cam.azimuth   = 145.0;
        recorder.cam.elevation = -22.0;
        recorder.cam.distance  = 1.35;
        recorder.cam.lookat[0] = 0.05;
        recorder.cam.lookat[1] = 0.02;
        recorder.cam.lookat[2] = 0.88;
        recording              = true;
    }

    mj_kdl::reset(&env);
    if (!headless && !mj_kdl::open_viewer(&env)) return 1;

    int        sim_step    = 0;
    bool       record_fail = false;
    const bool completed   = ex::run_phases(
      env,
      { { &robot, fingers } },
      phases,
      restart,
      [&](std::size_t, const KDL::JntArray &q_des) { ex::pd_gravity(robot, dyn, q_des, kKp, kKd); },
      [&](const ex::Phase &, double) {
          if (recording && ++sim_step % steps_per_frame == 0
              && !mj_kdl::record_frame(&recorder, &env)) {
              std::cerr << "record_frame() failed at step " << sim_step << "\n";
              mj_kdl::cleanup(&recorder);
              recording   = false;
              record_fail = true;
          }
      }
    );
    mj_kdl::update(&env);
    if (recording) {
        mj_kdl::cleanup(&recorder);
        std::cout << "Saved recording: " << args.record_path << "\n";
    }

    int         in_receiver = 0;
    KDL::Vector centroid    = KDL::Vector::Zero();
    for (const auto *g : grains) {
        if (inside_receiver(g->pose.p)) ++in_receiver;
        centroid += g->pose.p / kNumBalls;
    }
    std::cout << "balls in transparent receiver: " << in_receiver << "/" << kNumBalls
              << " (at least " << kMinInReceiver << ")\n"
              << std::fixed << std::setprecision(3) << "grain centroid: [" << centroid.x() << ", "
              << centroid.y() << ", " << centroid.z() << "] receiver center=[" << kJugX << ", "
              << kJugY << "]\n";
    mj_kdl::cleanup(&env);
    const bool ok = completed && !record_fail && in_receiver >= kMinInReceiver;
    return headless ? ex::verdict(ok, "the balls were poured into the receiver") : 0;
}

File: python/examples/ex_table_pour.py
#!/usr/bin/env python3
"""Table pour example ported from src/examples/ex_table_pour.cpp."""

from __future__ import annotations

import argparse
import math
import os
from dataclasses import dataclass
from pathlib import Path

import PyKDL as kdl
import mj_kdl_wrapper as mjk

ARM = "third_party/menagerie/kinova_gen3/gen3.xml"
GRIPPER = "assets/robotiq_2f85/2f85.xml"
TABLE = "assets/table.xml"
BOTTLE = "assets/mug.xml"
RECEIVER = "assets/mug_table.xml"
HOME = [0.0, 0.2618, 3.1416, -2.2689, 0.0, 0.9599, 1.5708]
TABLE_Z = 0.70
ROBOT_BACK_X = -0.26
JUG_X = 0.30
JUG_Y = 0.14
RETREAT_X = JUG_X - 0.08
RETREAT_Y = JUG_Y - 0.08
BALL_RADIUS = 0.007
NUM_BALLS = 36
POUR_TILT_RAD = 1.95
IK_TOL = 3e-3
KP = [120.0, 220.0, 120.0, 220.0, 110.0, 190.0, 90.0]
KD = [12.0, 22.0, 12.0, 22.0, 11.0, 18.0, 9.0]


@dataclass(frozen=True)
class Phase:
    name: str
    target: list[float]
    duration: float
    timeout: float
    settle_tol: float
    gripper: float


class ResetRequested(Exception):
    """Raised when the simulate UI reset is detected, to restart the sequence."""


def path(value: str, label: str) -> Path:
    p = Path(value)
    if not p.exists():
        raise FileNotFoundError(f"{p} does not exist for {label}; run from the repo root")
    return p


def gripper_attachment(gripper_path: Path) -> mjk.AttachmentSpec:
    attach = mjk.AttachmentSpec()
    attach.mjcf_path = str(gripper_path)
    attach.attach_to = mjk.AttachTarget(mjk.AttachKind.Site, "pinch_site")
    attach.prefix = "g_"
    return attach


def bottle_attachment(bottle_path: Path) -> mjk.AttachmentSpec:
    attach = mjk.AttachmentSpec()
    attach.mjcf_path = str(bottle_path)
    attach.attach_to = mjk.AttachTarget(mjk.AttachKind.Body, "g_base")
    attach.prefix = "pour_"
    return attach


def table_object(table_path: Path) -> mjk.SceneObject:
    table = mjk.SceneObject()
    table.name = "table"
    table.mjcf_path = str(table_path)
    table.pos = [0.0, 0.0, TABLE_Z]
    table.fixed = True
    return table


def receiver_object(receiver_path: Path) -> mjk.SceneObject:
    recv = mjk.SceneObject()
    recv.name = "recv"
    recv.mjcf_path = str(receiver_path)
    recv.pos = [JUG_X, JUG_Y, TABLE_Z]
    return recv


def ball_object(index: int) -> mjk.SceneObject:
    ball = mjk.SceneObject()
    ball.name = f"grain_{index:02d}"
    ball.shape = mjk.Shape.SPHERE
    ball.size = [BALL_RADIUS, 0.0, 0.0]
    ball.pos = [0.0, 0.0, TABLE_Z + 0.40 + index * 2.0 * BALL_RADIUS]
    ball.rgba = [1.0, 0.84, 0.30, 1.0]
    ball.mass = 0.006
    ball.condim = mjk.Condim.Torsional
    ball.friction = [0.5, 0.02, 0.001]
    return ball


def build_env() -> tuple[mjk.Env, mjk.Robot]:
    spec = mjk.SceneSpec()
    spec.timestep = 0.002
    spec.add_floor = True
    spec.add_skybox = True
    spec.objects = [
        table_object(path(os.environ.get("MJ_KDL_TABLE", TABLE), "table model")),
        *[ball_object(i) for i in range(NUM_BALLS)],
        receiver_object(path(os.environ.get("MJ_KDL_RECEIVER", RECEIVER), "receiver model")),
    ]

    robot_spec = mjk.RobotSpec()
    robot_spec.path = str(path(mjk.menagerie.model_path("kinova_gen3", env_var="MJ_KDL_MODEL"), "arm model"))
    robot_spec.pos = [ROBOT_BACK_X, 0.0, TABLE_Z]
    robot_spec.attachments = [
        gripper_attachment(path(os.environ.get("MJ_KDL_GRIPPER", GRIPPER), "gripper model")),
        bottle_attachment(path(os.environ.get("MJ_KDL_BOTTLE", BOTTLE), "bottle model")),
    ]
    spec.robots = [robot_spec]

    env = mjk.Env.build(spec)
    tool = mjk.ToolFrameSpec()
    tool.tool_body = "g_base"
    tool.tcp_site = "g_pinch"
    robot = env.create_robot("base_link", "bracelet_link", tool=tool)
    return env, robot


def jnt(values: list[float]) -> kdl.JntArray:
    q = kdl.JntArray(len(values))
    for i, value in enumerate(values):
        q[i] = value
    return q


def as_list(q: kdl.JntArray) -> list[float]:
    return [q[i] for i in range(q.rows())]


def clamp(value: float, low: float, high: float) -> float:
    return max(low, min(high, value))


def clamp_joint(value: float, limit: tuple[float, float]) -> float:
    low, high = limit
    if math.isfinite(low) and math.isfinite(high) and high > low:
        return clamp(value, low, high)
    return value


def joint_limit_arrays(robot: mjk.Robot) -> tuple[kdl.JntArray, kdl.JntArray]:
    q_min = kdl.JntArray(robot.n_joints)
    q_max = kdl.JntArray(robot.n_joints)
    for i, (low, high) in enumerate(robot.joint_limits):
        if math.isfinite(low) and math.isfinite(high) and high > low:
            q_min[i], q_max[i] = low, high
        else:
            q_min[i], q_max[i] = -2.0 * math.pi, 2.0 * math.pi
    return q_min, q_max


def build_waypoints(env: mjk.Env, robot: mjk.Robot) -> dict[str, list[float]]:
    chain = robot.kdl_chain()
    n = robot.n_joints
    fk = kdl.ChainFkSolverPos_recursive(chain)
    q_min, q_max = joint_limit_arrays(robot)
    ik_vel = kdl.ChainIkSolverVel_pinv(chain)
    ik_nr = kdl.ChainIkSolverPos_NR_JL(chain, q_min, q_max, fk, ik_vel, 2000, 1e-5)
    ik_lma = kdl.ChainIkSolverPos_LMA(chain, 1e-5, 2000)

    q_home = jnt(HOME)
    home_fk = kdl.Frame()
    fk.JntToCart(q_home, home_fk)
    # Carry the bottle at the home orientation, tilted slightly forward.
    carry_tcp = home_fk.M * kdl.Rotation.RotY(-0.05)

    world_T_base = kdl.Frame(kdl.Rotation.Identity(), kdl.Vector(ROBOT_BACK_X, 0.0, TABLE_Z))
    base_T_world = world_T_base.Inverse()

    # Constant TCP->outlet offset, measured at the live home configuration.
    robot.set_joint_pos(HOME, call_forward=False)
    world_T_outlet = env.site_frame("pour_outlet")
    world_T_tcp = env.site_frame("g_pinch")
    tcp_outlet = world_T_tcp.Inverse() * world_T_outlet.p

    def outlet_target_to_tcp_target(tcp_rot: kdl.Rotation, outlet_pos: kdl.Vector) -> kdl.Frame:
        return kdl.Frame(tcp_rot, outlet_pos - tcp_rot * tcp_outlet)

    def solve(name: str, seed_values: list[float], outlet_pos: kdl.Vector) -> list[float]:
        target = base_T_world * outlet_target_to_tcp_target(carry_tcp, outlet_pos)
        seed = jnt(seed_values)
        out = kdl.JntArray(n)
        ok = ik_nr.CartToJnt(seed, target, out) >= 0
        if not ok:
            ok = ik_lma.CartToJnt(seed, target, out) >= 0
        if not ok:
            raise RuntimeError(f"IK failed for {name}")
        fk_out = kdl.Frame()
        fk.JntToCart(out, fk_out)
        if (target.p - fk_out.p).Norm() > IK_TOL:
            raise RuntimeError(f"IK pose error for {name}")
        return as_list(out)

    q_pre_pour = solve("pre-pour", HOME, kdl.Vector(JUG_X, JUG_Y, TABLE_Z + 0.27))
    q_pour = solve("pour", q_pre_pour, kdl.Vector(JUG_X, JUG_Y, TABLE_Z + 0.20))
    q_retreat = solve("retreat", q_pour, kdl.Vector(RETREAT_X, RETREAT_Y, TABLE_Z + 0.27))
    q_tilt = q_pour[:]
    q_tilt[-1] = clamp_joint(q_tilt[-1] + POUR_TILT_RAD, robot.joint_limits[-1])
    return {
        "home": HOME[:],
        "pre_pour": q_pre_pour,
        "pour": q_pour,
        "tilt": q_tilt,
        "retreat": q_retreat,
    }


def apply_pd_gravity(robot: mjk.Robot, target: list[float]) -> None:
    robot.update()
    gravity = robot.gravity_torques(-9.81)
    robot.jnt_trq_cmd = [
        KP[i] * (target[i] - robot.jnt_pos_msr[i]) - KD[i] * robot.jnt_vel_msr[i] + gravity[i]
        for i in range(robot.n_joints)
    ]


def max_abs_joint_err(robot: mjk.Robot, target: list[float]) -> float:
    return max(abs(target[i] - robot.jnt_pos_msr[i]) for i in range(robot.n_joints))


def lerp(start: list[float], target: list[float], alpha: float) -> list[float]:
    return [a + alpha * (b - a) for a, b in zip(start, target)]


def place_balls_in_bottle(env: mjk.Env, robot: mjk.Robot) -> None:
    robot.set_joint_pos(HOME, call_forward=True)
    center = env.site_frame("pour_center")
    spacing = 2.0 * BALL_RADIUS
    for i in range(NUM_BALLS):
        layer = i // 9
        slot = i % 9
        ix = float(slot % 3) - 1.0
        iy = float(slot // 3) - 1.0
        local = kdl.Vector(ix * spacing, iy * spacing, -0.026 + layer * spacing)
        world = center * local
        env.set_body_pose(f"grain_{i:02d}", [world.x(), world.y(), world.z()])


def balls_in_receiver(env: mjk.Env) -> tuple[int, list[float]]:
    count = 0
    centroid = [0.0, 0.0, 0.0]
    for i in range(NUM_BALLS):
        frame = env.body_frame(f"grain_{i:02d}")
        pos = [frame.p.x(), frame.p.y(), frame.p.z()]
        centroid = [a + b for a, b in zip(centroid, pos)]
        if (
            abs(pos[0] - JUG_X) < 0.040
            and abs(pos[1] - JUG_Y) < 0.040
            and TABLE_Z + 0.004 < pos[2] < TABLE_Z + 0.13
        ):
            count += 1
    return count, [value / NUM_BALLS for value in centroid]


def step_once(robot: mjk.Robot, viewer: mjk.SimulateViewer | None) -> bool:
    if viewer is not None:
        return viewer.step()
    return robot.step()


def run_phase(
    env: mjk.Env,
    robot: mjk.Robot,
    phase: Phase,
    viewer: mjk.SimulateViewer | None,
    recorder: mjk.VideoRecorder | None,
    record_every: int,
    step_counter: list[int],
    state: dict,
) -> bool:
    print(f"State: {phase.name}")
    robot.update()
    start = robot.jnt_pos_msr[:]
    t0 = env.time()
    while True:
        elapsed = env.time() - t0
        alpha = clamp(elapsed / phase.duration, 0.0, 1.0) if phase.duration > 0.0 else 1.0
        apply_pd_gravity(robot, lerp(start, phase.target, alpha))
        if env.has_actuator("g_fingers_actuator"):
            env.set_actuator_ctrl("g_fingers_actuator", phase.gripper)

        done_time = elapsed >= phase.duration
        done_pose = phase.settle_tol < 0.0 or max_abs_joint_err(robot, phase.target) <= phase.settle_tol
        done_timeout = phase.timeout > 0.0 and elapsed >= phase.timeout
        if (done_time and done_pose) or done_timeout:
            return True
        if viewer is not None and not viewer.is_running():
            return False
        if not step_once(robot, viewer):
            return False
        if viewer is not None and env.time() < state["prev"] - 1e-6:  # UI reset pressed
            env.reset()
            state["prev"] = env.time()
            raise ResetRequested()
        state["prev"] = env.time()
        step_counter[0] += 1
        if recorder is not None and step_counter[0] % record_every == 0:
            recorder.record_frame()


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--gui", action="store_true")
    parser.add_argument("--record", nargs="?", const="table_pour.mp4")
    parser.add_argument("--headless", action="store_true", help=argparse.SUPPRESS)
    args = parser.parse_args()

    env, robot = build_env()
    recorder = None
    try:
        robot.ctrl_mode = mjk.CtrlMode.TORQUE

        def on_reset(ctx):
            place_balls_in_bottle(env, robot)  # also re-homes the arm
            if env.has_actuator("g_fingers_actuator"):
                env.set_actuator_ctrl("g_fingers_actuator", 255.0)

        env.on_reset = on_reset
        env.reset()

        waypoints = build_waypoints(env, robot)
        phases = [
            Phase("HOME", waypoints["home"], 0.8, 2.0, 0.08, 255.0),
            Phase("PRE_POUR", waypoints["pre_pour"], 2.0, 4.0, 0.08, 255.0),
            Phase("POUR", waypoints["pour"], 1.8, 4.0, 0.07, 255.0),
            Phase("TILT", waypoints["tilt"], 3.0, 5.0, 0.07, 255.0),
            Phase("POUR_HOLD", waypoints["tilt"], 2.5 if not args.gui else 10.0, 0.0, -1.0, 255.0),
            Phase("RETREAT", waypoints["retreat"], 1.6, 3.0, 0.08, 255.0),
            Phase("HOLD", waypoints["retreat"], 1.0 if not args.gui else 10.0, 0.0, -1.0, 255.0),
        ]

        fps = 60
        record_every = max(1, int(1.0 / (fps * env.timestep())))
        if args.record:
            recorder = mjk.VideoRecorder.open_preset(
                env, args.record, mjk.VideoResolution.R1080p, fps
            )
        step_counter = [0]
        state = {"prev": env.time()}
        if args.gui:
            viewer = mjk.SimulateViewer.open(robot, "ex_table_pour.py")
            try:
                while viewer.is_running():
                    try:
                        for phase in phases:
                            if not run_phase(
                                env, robot, phase, viewer, recorder, record_every, step_counter, state
                            ):
                                raise StopIteration
                        break
                    except ResetRequested:
                        continue
                    except StopIteration:
                        break
            finally:
                viewer.close()
        else:
            for phase in phases:
                if not run_phase(
                    env, robot, phase, None, recorder, record_every, step_counter, state
                ):
                    break
        in_receiver, centroid = balls_in_receiver(env)
        print(f"balls in transparent receiver: {in_receiver}/{NUM_BALLS}")
        print(f"grain centroid: {[round(v, 3) for v in centroid]} receiver center={[JUG_X, JUG_Y]}")
        if recorder is not None:
            print(f"recorded: {args.record}")
    finally:
        if recorder is not None:
            recorder.close()
        env.close()
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
